### ISSUE TITLE
PIRLS step-halving failure in Gamma GLMM

### ISSUE THREAD

**Issue by lpitombo** [lme4/lme4#179]
Dear authors,

I'm exploring the glmer function in lme4. I had some troubles when I exclude/include some factors. I runned 5 models with diferent combinations of factors, but in the model 6, which is the same than model 3 but with 1 less factor, it gave the message:

Error: (maxstephalfit) PIRLS step-halvings failed to reduce deviance in pwrssUpdate

here are the codes: 

> microbial1<-glmer(N2O~organicC+fertilization+straw+NH4+NO3+rel_Bacilli+rel_Cloroflexi+(1|day), family=Gamma, na.action = na.omit)
> microbial2<-glmer(N2O~organicC+fertilization+straw+NH4+NO3+rel_Lactobacillales+rel_Cloroflexi+(1|day), family=Gamma, na.action = na.omit)
> microbial3<-glmer(N2O~organicC+fertilization+straw+NH4+NO3+rel_Bacillales+rel_Cloroflexi+(1|day), family=Gamma, na.action = na.omit)
> microbial4<-glmer(N2O~organicC+fertilization+straw+NH4+NO3+(1|day), family=Gamma, na.action = na.omit)
> microbial5<-glmer(N2O~organicC+fertilization+straw+NH4+NO3+rel_Cloroflexi+(1|day), family=Gamma, na.action = na.omit)
> microbial6<-glmer(N2O~organicC+fertilization+straw+NH4+NO3+rel_Bacilli+(1|day), family=Gamma, na.action = na.omit)
> microbial6<-glmer(N2O~organicC+fertilization+straw+NH4+NO3+rel_Bacillales+(1|day), family=Gamma, na.action = na.omit)
> Error: (maxstephalfit) PIRLS step-halvings failed to reduce deviance in pwrssUpdate

I would be grateful if someone can help me.

Regards,
Leonardo

**Comment by RocioJara** [comment]
Was this issue ever solved? I have a similar problem but with binomial family.

Thanks,
Rocio

**Comment by bbolker** [comment]
Can you give more detail/a reproducible example?  By "similar problem" do you mean you're getting PIRLS-related errors, or ... ? What link function are you using? PIRLS failures occur pretty rarely (one of the reason we haven't figured everything out yet), and [while this sounds like a big cop-out ...] are usually associated with wonky data sets/very unstable fits ...

**Comment by RocioJara** [comment]
Sorry for my incomplete message. Here is some more information: 

I am using the glmer function in lme4  (package version 1.1.18.1). I am using this mixed model to estimate  Daily Survival Rate model using a custom link function (logistic-exposure). I have several variables, but only when trying to include one of them (height), independent of the order it enters the model, and if it is by itself or along other variables, I will get the error: "Error in pwrssUpdate: PIRLS step-halvings failed to reduce deviance in pwrssUpdate".  Here is an example of a line that would not run 

global.chincol <- glmer(Fate ~  camera + nest.age + height + (1 | year), family = binomial(link = logexp(DSR.chincol$Exposure)), data = DSR.chincol, na.action = "na.fail"))

And this is how my link function is set, just in case 

library(MASS)
logexp <- function(exposure = 1)
{
        linkfun <- function(mu) qlogis(mu^(1/exposure))
        linkinv <- function(eta)  plogis(eta)^exposure
        logit_mu_eta <- function(eta) {
                ifelse(abs(eta)>30,.Machine$double.eps,
                       exp(eta)/(1+exp(eta))^2)
        }
        mu.eta <- function(eta) {       
                exposure * plogis(eta)^(exposure-1) *
                        logit_mu_eta(eta)
        }
        valideta <- function(eta) TRUE
        link <- paste("logexp(", deparse(substitute(exposure)), ")",
                      sep="")
        structure(list(linkfun = linkfun, linkinv = linkinv,
                       mu.eta = mu.eta, valideta = valideta, 
                       name = link),
                  class = "link-glm")
}

I am doing this analysis for different data sets, and so far I am encountering this problem with only one data set. This means that for other data set I can include "height" in my models without problems (all data sets have exactly the same structure). That makes me agree with you when you say this errors are usually associated with wonky data sets/very unstable fits. So I would like to know how to deal with this in this particular data set, as the inclusion of that variable is important.

Unfortunately I can't provide the data here, but I would be so grateful if you could take a look at it. Can I send it to you if you keep it confidential until publication? If yes, which email should I use? 

Thanks
Rocio

**Comment by bbolker** [comment]
Please send to the maintainer e-mail, `bbolker+lme4@gmail.com`: this will also go to the two co-authors (Bates & Mächler), but I can vouch for their keeping the data confidential.

It's slightly reassuring/less surprising that you're having trouble when using an unusual (custom) link. (As an aside, it would make life a little easier if people used log-hazard models rather than power-exponential models, but I guess an opportunity to break the model is also an opportunity to make it more robust ...)

**Comment by bbolker** [comment]
You have a [complete separation](http://bbolker.github.io/mixedmodels-misc/glmmFAQ.html#penalizationhandling-complete-separation) problem; there are only 5 or so unique values of `height`, responses in the first category have both 0 and 1 responses, responses in the other categories have all 1 responses. See the link for potential solutions ...

**Comment by RocioJara** [comment]
Thank you so much for your help!

**Comment by RocioJara** [comment]
I have taken care of the complete separation problem. However, I am getting the same error message when trying to run exactly the same code for other species. I checked the variables that are giving me problems and there is no complete separation this time. I would appreciate it so much if you could  take a look a this again

**Comment by RocioJara** [comment]
Yes, I did send that data a week ago. More than a reminder, I realized it may have been inappropriate to contact you to the email directly, so I decided to post the question here instead.

**Comment by bbolker** [comment]
I hate to be pessimistic, but when I look at the raw numbers of failures and successes in your data:
```
            0   1
cachudito   4 174
zorzal      4  67
Cometocino  6 280
chincol    20 322
fio        18 277
```

... I'm amazed you can fit anything at all, let alone this many parameters with a custom link. A good rule of thumb for regression is that you need *at least* 10 "effective" samples per parameter you try to estimate, where "effective" for binary outcomes is the *minimum* of the number of outcomes=0 and the number of outcomes=1. (This is from Chapter 4 of Frank Harrell's *Regression Modeling Strategies* book: basic info is listed [here](http://wiki.math.yorku.ca/index.php/Statistics:_Sample_size_-_validity), more discussion [here](https://stats.stackexchange.com/questions/286521/binomial-mixed-effect-model-adjustments-and-limitations-thereof/286523#286523).) That means that for your first three species I wouldn't try to estimate more than 1 parameter, for your other two 2 parameters at most. I just don't think you have the sample size here (no matter how difficult it has been to get these data!) to do anything more than some very basic descriptive statistics, maybe a univariate fit or two on the most interesting parameters ...

```
spp <- c("cachudito","zorzal","Cometocino","chincol","fio")
tabfun <- function(x) {
    ff <- list.files(pattern=sprintf("DSR[_.]%s.csv",x))
    xx <- read.csv(ff)
    c(table(xx$Fate))
}
res <- do.call(rbind,lapply(spp,tabfun))
rownames(res) <- spp
```

**Comment by bbolker** [comment]
Yes. The issue is not that there anything intrinsically incorrect about a badly overparameterized/underpowered model (note that [as Georges Monette says in the first page linked above] the recommendations for minimal n/p will vary by subject area depending on the typical signal-to-noise ratio; Harrell's are for clinical biostatistics, but I see no reason to think that signal-to-noise ratio will be much better in ecology and evolution ...)  The issue is at a higher level - underpowered models will perform poorly (wide confidence intervals, high p-values, low adjusted R^2, poor predictive ability ...), which in turn will lead to disappointment, temptation to p-hack/snoop/etc. ...  I really do recommend reading Harrell's chapter 4 for the whole story ...

**Comment by bbolker** [comment]
Another philosophical point: *in principle* (in a perfect world) GLMMs should be able to fit these models, just like (for example) you can compute an estimate of the variance from two data points ... and it's true that the effort invested to make marginally sensible models like this work would improve the stability/robustness of the code overall. But it's hard, and not obvious that it's the most valuable place to invest effort. (FWIW I tried glmmTMB (with a cloglog link and a log-exposure offset, since the power-logistic link isn't implemented) on one of these problems and got some sort of equivalent error (I think it was NA/NaN in the objective function).

**Comment by Jailos** [comment]
M2 <- glmer(Cases_U5~offset(log(Population_Under5))+LLIN+IRS+Mean.Max.Temp+Mean.Min.Temp+Mean.Temp+Pop_density.Km2+avg_Elevation+NDVI+xMean.mm.day+xMax.mm.day+(1|area_name), data = mydata2, negative.binomial(2)) 

I have a challenge fitting glmer "poisson" negative.binomial(2) as shown above using my data getting a series of error warnings even after a series of troubleshooting and they are:

1.![image](https://user-images.githubusercontent.com/7884680/55631132-37e29400-57af-11e9-82a3-e6b7f463dbbe.png)


*so I rescaled the data using #mydata2[pvars] <- lapply(mydata2[pvars],scale)

When I ran the code again, I the first part of the previous error message was still there:
2. 
![image](https://user-images.githubusercontent.com/7884680/55631221-74ae8b00-57af-11e9-9b05-0ee0b545aec8.png)

when I changed to use negative.binomial(2), the error changes as below
![image](https://user-images.githubusercontent.com/7884680/55631288-990a6780-57af-11e9-85fb-80146db2b861.png)

I have tried many things as suggested on forums but nothing is working like adding control
3. 
![image](https://user-images.githubusercontent.com/7884680/55631805-c99ed100-57b0-11e9-95fc-e03838a036ec.png)

Nothing seems to work. 
Please help.

**Comment by bbolker** [comment]
Don't know if @Jailos ever got an answer, this seems to have slipped through the cracks.  Can we have a reproducible example?

**Comment by N-Accumbens** [comment]
I am having the same (?) issue and I have a reproducible example. I gather that this issue is rare, so hopefully my reprex can be helpful, although I'm unable to follow the statistical details.

``` r
library(lme4)
#> Loading required package: Matrix
#> Warning: package 'Matrix' was built under R version 4.0.5
df.odd <- data.frame(
  Subject = rep(seq(1, 24, 2), each = 4),
  Condition = c("A", "B", "A", "B"),
  Block = c(1, 2, 3, 4),
  Yes = sample(1:10, 24, replace = TRUE)
)
df.odd$No <- (10 - df.odd$Yes)

df.even <- data.frame(Subject = rep(seq(2, 24, 2), each = 4),
  Condition = c("A", "B", "A", "B"),
  Block = c(1, 2, 3, 4),
  Yes = sample(1:10, 24, replace = TRUE))

df.even$No <- (10 - df.even$Yes)

df <- rbind(df.odd, df.even)

df$Counterbalance <- df$Subject %% 2

glmer(formula = cbind(Yes, No) ~ Condition + Block + Counterbalance + (1|Subject),
      family = binomial(link = "log"),
      data=df)
#> Error in (function (fr, X, reTrms, family, nAGQ = 1L, verbose = 0L, maxit = 100L, : (maxstephalfit) PIRLS step-halvings failed to reduce deviance in pwrssUpdate
```

<sup>Created on 2021-07-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

This is my sessionInfo(), if you are interested
``` r
library(lme4)
#> Loading required package: Matrix
#> Warning: package 'Matrix' was built under R version 4.0.5
sessionInfo()
#> R version 4.0.4 (2021-02-15)
#> Platform: x86_64-w64-mingw32/x64 (64-bit)
#> Running under: Windows 10 x64 (build 19043)
#> 
#> Matrix products: default
#> 
#> locale:
#> [1] LC_COLLATE=English_United States.1252 
#> [2] LC_CTYPE=English_United States.1252   
#> [3] LC_MONETARY=English_United States.1252
#> [4] LC_NUMERIC=C                          
#> [5] LC_TIME=English_United States.1252    
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] lme4_1.1-26  Matrix_1.3-3
#> 
#> loaded via a namespace (and not attached):
#>  [1] Rcpp_1.0.6        rstudioapi_0.13   knitr_1.33        magrittr_2.0.1   
#>  [5] splines_4.0.4     MASS_7.3-54       statmod_1.4.36    lattice_0.20-44  
#>  [9] rlang_0.4.11      minqa_1.2.4       stringr_1.4.0     highr_0.9        
#> [13] tools_4.0.4       grid_4.0.4        nlme_3.1-152      xfun_0.22        
#> [17] cli_2.5.0         withr_2.4.2       htmltools_0.5.1.1 yaml_2.2.1       
#> [21] digest_0.6.27     nloptr_1.2.2.2    fs_1.5.0          ps_1.6.0         
#> [25] glue_1.4.2        evaluate_0.14     rmarkdown_2.8     reprex_2.0.0     
#> [29] stringi_1.6.1     compiler_4.0.4    boot_1.3-28
```

<sup>Created on 2021-07-15 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

In my context, I have participants learning 4 different list of words and are tested 4 times, once for each list. During the test time, they are asked to say as many words as they can recall for the list they just learned. The number of words they recall is denoted in "Yes", number of words they don't recall in "No". I am interested in comparing "A" and "B". Order of conditions is counterbalanced based on subject number.

I heard that the `glmer` code above might help me analyze the data, but it doesn't seem to run. Hopefully this was helpful enough.
